### PR TITLE
Fix crash in ``super-init-not-called`` checker

### LINKED ISSUE #6027: '_CData' not found in <ClassDef.Union l.None at 0x7fb308fdb650>.
### Bug description

internal crash when doing a pylint run

full details here: 

https://github.com/KNX-IOT/KNX-IOT-STACK/runs/5733695443?check_suite_focus=true

### Configuration

```ini
blank
```

### Command used

```shell
see github action
```

### Pylint output

```shell
pylint crashed with a ``NameInferenceError`` and with the following stacktrace:

Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pylint/lint/pylinter.py", line 1111, in _check_files
    self._check_file(get_ast, check_astroid_module, file)
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pylint/lint/pylinter.py", line 1146, in _check_file
    check_astroid_module(ast_node)
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pylint/lint/pylinter.py", line 1299, in check_astroid_module
    ast_node, walker, rawcheckers, tokencheckers
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pylint/lint/pylinter.py", line 1345, in _check_astroid_module
    walker.walk(node)
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pylint/utils/ast_walker.py", line 76, in walk
    self.walk(child)
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pylint/utils/ast_walker.py", line 76, in walk
    self.walk(child)
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pylint/utils/ast_walker.py", line 73, in walk
    callback(astroid)
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pylint/checkers/classes/class_checker.py", line 1095, in visit_functiondef
    self._check_init(node, klass)
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/pylint/checkers/classes/class_checker.py", line 2004, in _check_init
    for inf_base in base.infer():
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/astroid/nodes/node_ng.py", line 168, in infer
    yield from self._infer(context=context, **kwargs)
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/astroid/decorators.py", line 136, in raise_if_nothing_inferred
    yield next(generator)
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/astroid/decorators.py", line 105, in wrapped
    for res in _func(node, context, **kwargs):
  File "/opt/hostedtoolcache/Python/3.7.12/x64/lib/python3.7/site-packages/astroid/inference.py", line 213, in infer_name
    name=self.name, scope=self.scope(), context=context
astroid.exceptions.NameInferenceError: '_CData' not found in <ClassDef.Union l.None at 0x7fb308fdb650>.
```

### Expected behavior

no crash

### Pylint version

```shell
python 3.7 (but also on others)
```

### OS / Environment

Linux (on github)

### Additional dependencies

see github action.

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #6027.

Should have never put that comment there...

### CODE REVIEW

**Comment by DanielNoord** [comment]
Uhm.. Does `pypy` support `types`?

File: pylint/checkers/classes/class_checker.py [hunk 1]
```diff
@@ -1999,11 +1999,12 @@ def _check_init(self, node: nodes.FunctionDef, klass_node: nodes.ClassDef) -> No
 
             # Return if any of the klass' first-order bases is protocol
             for base in klass.bases:
-                # We don't need to catch InferenceError here as _ancestors_to_call
-                # already does this for us.
-                for inf_base in base.infer():
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/classes/class_checker.py:2004]: Should we use ``safe_infer`` and a ``is not None`` check instead ? if

**Review by Pierre-Sassoulas — COMMENTED** [review]
Maybe we need to forbid the use of infer accross all the  pylint codebase, there's no way we're not going to get a crash eventuellay with the diversity of code pylint is going to be run on.

File: pylint/checkers/classes/class_checker.py (hunk 1 above)
Comment (DanielNoord) [reply pylint/checkers/classes/class_checker.py:2004]: The problem of using `safe_infer` is that it only returns one node. Sometimes I don't care if there is ambiguity about the nodes, if I can recognise one of them as being a typing protocol (as I do in this case) I want to exit. `safe_infer` doesn't really allow that.

Comment (Pierre-Sassoulas) [reply pylint/checkers/classes/class_checker.py:2004]: Ho I thought it was equivalent but with all exceptions caught, thanks..

**Review by Pierre-Sassoulas — APPROVED** [review]

File: ChangeLog [hunk 2]
```diff
@@ -41,6 +41,10 @@ Release date: TBA
 
   Closes #6028
 
+* Fix crash in ``super-init-not-called`` checker.
```
Comment (Pierre-Sassoulas) [inline ChangeLog:44]: Telling in what case there was a crash is classier
.
```suggestion
* Fix crash in ``super-init-not-called`` checker when using ``ctypes.Union``.
```

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by Pierre-Sassoulas** [comment]
We might need to do 
```
[testoptions]
except_implementations=PyPy
```
What do you think @DanielNoord ?

**Comment by DanielNoord** [comment]
> We might need to do
> 
> ```
> [testoptions]
> except_implementations=PyPy
> ```
> 
> What do you think @DanielNoord ?

I mean we could, but I don't know why this doesn't work on PyPy. I don't know the `ctypes` module...

**Comment by jacobtylerwalls** [comment]
Looks like the PyPy behavior is correct -- super's init is not being called, so the test file should have a `# [super-init-not-called]`, but because astroid was raising InferenceError, it wasn't working for CPython -- which is good for our present purposes -- we want to test that no crash happens on an inference failure. So maybe a second file with the correct message for PyPy would do it?

**Comment by DanielNoord** [comment]
I guess raising if we can't infer will lead to more issues right?

File: pylint/checkers/classes/class_checker.py [hunk 3]
```diff
@@ -1999,11 +1999,12 @@ def _check_init(self, node: nodes.FunctionDef, klass_node: nodes.ClassDef) -> No
...
-                    if inf_base.qname() in utils.TYPING_PROTOCOLS:
-                        return
+                try:
+                    for inf_base in base.infer():
+                        if inf_base.qname() in utils.TYPING_PROTOCOLS:
+                            return
+                except astroid.InferenceError:
+                    return
```
Comment (jacobtylerwalls) [inline pylint/checkers/classes/class_checker.py:2007]: Why not `continue`?

**Comment by Pierre-Sassoulas** [comment]
> I guess raising if we can't infer will lead to more issues right?

Yes I think we should prefer a false negative always when we're not sure.

File: doc/whatsnew/2.13.rst [hunk 4]
```diff
@@ -575,3 +575,7 @@ Other Changes
   separated (e.g. by parens).
 
   Closes #5769
+
+* Fix crash in ``super-init-not-called`` checker when using ``ctypes.Union``.
+
+  Closes #6027
```
Comment (jacobtylerwalls) [inline doc/whatsnew/2.13.rst:581]: Per https://github.com/PyCQA/pylint/pull/6040#discussion_r837901916 I think we're not adding hotfixes to the whatsnew -- just the changelog.

```suggestion
```

**Pull request merged**